diff --git a/libmamba/CMakeLists.txt b/libmamba/CMakeLists.txt
index 5fe133c..5aebce1 100644
--- a/libmamba/CMakeLists.txt
+++ b/libmamba/CMakeLists.txt
@@ -552,7 +552,7 @@ macro(libmamba_create_target target_name linkage output_name)
             find_library(LIBLZMA_LIBRARIES lzma REQUIRED)
             find_library(LZ4_LIBRARY NAMES lz4)
             find_library(LZO2_LIBRARY NAMES lzo2)
-            find_package(zstd CONFIG REQUIRED)
+            find_library(ZSTD_LIBRARY NAMES zstd REQUIRED)
             find_library(BZIP2_LIBRARIES NAMES bz2)
             find_library(CRYPTO_LIBRARIES NAMES libcrypto)
 
@@ -569,7 +569,7 @@ macro(libmamba_create_target target_name linkage output_name)
                     ${LIBXML2_LIBRARY}
                     ${ICONV_LIBRARY}
                     ${CHARSET_LIBRARY}
-                    zstd::libzstd_static
+                    ${ZSTD_LIBRARY}
                     ${LZ4_LIBRARY}
                     ${LZO2_LIBRARY}
                     ${BZIP2_LIBRARIES}
@@ -589,7 +589,7 @@ macro(libmamba_create_target target_name linkage output_name)
 
         find_package(CURL REQUIRED)
         find_package(LibArchive REQUIRED)
-        find_package(zstd REQUIRED)
+        find_library(ZSTD_LIBRARIES NAMES zstd)
         find_package(BZip2 REQUIRED)
         find_package(OpenSSL REQUIRED)
 
@@ -612,7 +612,7 @@ macro(libmamba_create_target target_name linkage output_name)
                 reproc
                 reproc++
                 simdjson::simdjson
-                zstd::libzstd_shared
+                ${ZSTD_LIBRARIES}
                 solv::libsolv
                 solv::libsolvext
                 solv::cpp
diff --git a/micromamba/CMakeLists.txt b/micromamba/CMakeLists.txt
index a75d436d..5d21e7b5 100644
--- a/micromamba/CMakeLists.txt
+++ b/micromamba/CMakeLists.txt
@@ -115,7 +115,7 @@ if(BUILD_SHARED AND UNIX)
 
     install(
         FILES ${CMAKE_CURRENT_BINARY_DIR}/etc/profile.d/mamba.sh
-        DESTINATION ${CMAKE_INSTALL_PREFIX}/etc/profile.d/
+        DESTINATION /etc/profile.d/
     )
 endif()
 

