diff --git a/libmamba/CMakeLists.txt b/libmamba/CMakeLists.txt
index 42cb1670..3be90987 100644
--- a/libmamba/CMakeLists.txt
+++ b/libmamba/CMakeLists.txt
@@ -511,7 +511,7 @@ macro(libmamba_create_target target_name linkage output_name)
     )

     if (WIN32)
-        find_path(WINREG_INCLUDE_DIR NAMES "winreg/WinReg.hpp")
+        find_path(WINREG_INCLUDE_DIR NAMES WinReg.hpp)
     endif()

     target_include_directories(
diff --git a/libmamba/src/core/shell_init.cpp b/libmamba/src/core/shell_init.cpp
index 83854582..3e7c920f 100644
--- a/libmamba/src/core/shell_init.cpp
+++ b/libmamba/src/core/shell_init.cpp
@@ -12,7 +12,7 @@
 #include <fmt/ostream.h>
 #include <reproc++/run.hpp>
 #ifdef _WIN32
-#include "winreg/WinReg.hpp"
+#include <WinReg.hpp>
 #endif

 #include "mamba/core/activation.hpp"
diff --git a/libmamba/src/core/util_os.cpp b/libmamba/src/core/util_os.cpp
index 21afe9c2..513da964 100644
--- a/libmamba/src/core/util_os.cpp
+++ b/libmamba/src/core/util_os.cpp
@@ -21,7 +21,7 @@
 // Incomplete header included last
 #include <tlhelp32.h>

-#include "winreg/WinReg.hpp"
+#include <WinReg.hpp>
 #endif

 #include <fmt/color.h>
